### PR TITLE
feat: upgrade moengage sdk version to latest v9.18.0

### PR DESCRIPTION
**Pull request by shwetalodhe123**
# Description

- Separated standard and non-standard traits keys.
- Simplified how standard trait keys are handled. They require a string-type value.
- Non-Standard trait keys could ideally take any values, as the MoEngage does not define them.
- Changed the method name of handleCustomAttribute to handleDateAndCustomUserAttribute. For better readability.
- Removed the identifyDateUserAttribute, as it is not used.
- Refactored the logic of the data handling method in handleDateAndCustomUserAttribute().

Type of change

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)

Checklist:

- [ ]  Version updated in README
- [ ]  Version updated in RudderSingular.xcodeproj
- [ ]  Version updated in RudderSingular.podspec
- [ ]  CHANGELOG Updated
- [ ]  I have performed a self-review of my own code
- [ ]  I have commented my code, particularly in hard-to-understand areas
- [ ]  I have added unit tests for the code
- [ ]  I have made corresponding changes to the documentation
- [ ]  Passed pod lib lint --no-clean --allow-warnings

### CODE REVIEW

File: Sources/Classes/RSMoEngageDestination.swift [hunk 1]
```diff
@@ -6,62 +6,83 @@
...
+    
+
 
     func track(message: TrackMessage) -> TrackMessage? {
         if !message.event.isEmpty {
             switch message.event {
-            case RSEvents.LifeCycle.applicationInstalled: MoEngage.sharedInstance().appStatus(INSTALL)
-            case RSEvents.LifeCycle.applicationUpdated: MoEngage.sharedInstance().appStatus(UPDATE)
+            case RSEvents.LifeCycle.applicationInstalled: MoEngageAppStatus(rawValue: 0)
+            case RSEvents.LifeCycle.applicationUpdated: MoEngageAppStatus.update
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:82]: In this code, you're not setting the app status. `MoEngageAppStatus.update` doesn't seem to do anything.

Please find the relevant code snippet from the MoEngage doc.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 2]
```diff
@@ -137,15 +165,15 @@ extension RSMoEngageDestination {
...
-                case RSKeys.Identify.Traits.gender: MoEngage.sharedInstance().setUserAttribute(value, forKey: USER_ATTRIBUTE_USER_GENDER)
-                case RSKeys.Identify.Traits.birthday: identifyDateUserAttribute(value: value, key: USER_ATTRIBUTE_USER_BDAY)
-                case RSKeys.Identify.Traits.address: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.address)
-                case RSKeys.Identify.Traits.age: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.age)
+                case RSKeys.Identify.Traits.email: MoEngageSDKAnalytics.sharedInstance.setEmailID((value as? String)!)
+                case RSKeys.Identify.Traits.name: MoEngageSDKAnalytics.sharedInstance.setName((value as? String)!)
+                case RSKeys.Identify.Traits.phone: MoEngageSDKAnalytics.sharedInstance.setMobileNumber(value as! String)
+                case RSKeys.Identify.Traits.firstName: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: "USER_ATTRIBUTE_USER_FIRST_NAME") //USER_ATTRIBUTE_USER_FIRST_NAME
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:171]: Remove comment.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 3]
```diff
@@ -137,15 +165,15 @@ extension RSMoEngageDestination {
...
+                case RSKeys.Identify.Traits.lastName: MoEngageSDKAnalytics.sharedInstance.setLastName((value as? String)!)
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:172]: We should not abort the execution if the value is not present. This might crash the SDK. Please make the necessary changes and remove the use of `!`.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 4]
```diff
@@ -157,11 +185,11 @@ extension RSMoEngageDestination {
             // Verify if the value is of type Date or not
             // Track UserAttribute using Epoch value. Refer here: https://developers.moengage.com/hc/en-us/articles/4403905883796-Tracking-User-Attributes
             if let value = value as? String, let convertedDate = dateFrom(isoDateString: value) {
-                MoEngage.sharedInstance().setUserAttributeTimestamp(convertedDate.timeIntervalSince1970, forKey: key)
+                MoEngageSDKAnalytics.sharedInstance.setUserAttribute(convertedDate.timeIntervalSince1970, withAttributeName: key)
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:188]: Please find the correct method which sets the timestamp.

File: Podfile [hunk 5]
```diff
@@ -11,7 +11,7 @@ end
 target 'RudderMoEngage' do
     project 'RudderMoEngage.xcodeproj'
     shared_pods
-    pod 'MoEngage-iOS-SDK', '~> 6.1.0'
+    pod 'MoEngage-iOS-SDK', '~> 9.18.0'
```
Comment (1abhishekpandey) [inline Podfile:14]: Please remove this. It is not needed.

File: RudderMoEngage.podspec [hunk 6]
```diff
@@ -1,12 +1,12 @@
 Pod::Spec.new do |s|
     s.name             = 'RudderMoEngage'
-    s.version          = '1.0.0'
+    s.version          = '1.1.0'
```
Comment (1abhishekpandey) [inline RudderMoEngage.podspec:3]: Please bump the version to `2.0.0` in all the files including Changelog, as you've updated the MoEngage API which might be a breaking change.

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]

File: Sources/Classes/RSMoEngageDestination.swift [hunk 7]
```diff
@@ -6,62 +6,83 @@
...
         if let anonymousId = message.anonymousId {
-            MoEngage.sharedInstance().setUserAttribute(anonymousId, forKey: "anonymousId")
+          
+            MoEngageSDKAnalytics.sharedInstance.setUserAttribute(anonymousId, withAttributeName: "anonymousId")
         }
+        
         if let userId = message.userId {
-            MoEngage.sharedInstance().setUserAttribute(userId, forKey: USER_ATTRIBUTE_UNIQUE_ID)
-            MoEngage.sharedInstance().setUserUniqueID(userId)
+            MoEngageSDKAnalytics.sharedInstance.setUserAttribute(userId, withAttributeName: "userId")
```
[the remark this answers is not in the document]
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: Done

File: RudderMoEngage.podspec (hunk 6 above)
Comment (shwetalodhe123) [reply RudderMoEngage.podspec:3]: done

File: Podfile (hunk 5 above)
Comment (shwetalodhe123) [reply Podfile:14]: done

File: Sources/Classes/RSMoEngageDestination.swift [hunk 8]
```diff
@@ -6,62 +6,83 @@
...
+    
+    
+
 
     func track(message: TrackMessage) -> TrackMessage? {
         if !message.event.isEmpty {
             switch message.event {
-            case RSEvents.LifeCycle.applicationInstalled: MoEngage.sharedInstance().appStatus(INSTALL)
-            case RSEvents.LifeCycle.applicationUpdated: MoEngage.sharedInstance().appStatus(UPDATE)
+            case RSEvents.LifeCycle.applicationInstalled: MoEngageAppStatus(rawValue: 0)
```
[the remark this answers is not in the document]
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: Done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 2 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: Done

File: Sources/Classes/RSMoEngageDestination.swift [hunk 9]
```diff
@@ -137,15 +165,15 @@ extension RSMoEngageDestination {
...
-                case RSKeys.Identify.Traits.lastName: MoEngage.sharedInstance().setUserLastName(value as? String)
-                case RSKeys.Identify.Traits.gender: MoEngage.sharedInstance().setUserAttribute(value, forKey: USER_ATTRIBUTE_USER_GENDER)
-                case RSKeys.Identify.Traits.birthday: identifyDateUserAttribute(value: value, key: USER_ATTRIBUTE_USER_BDAY)
-                case RSKeys.Identify.Traits.address: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.address)
-                case RSKeys.Identify.Traits.age: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.age)
+                case RSKeys.Identify.Traits.email: MoEngageSDKAnalytics.sharedInstance.setEmailID((value as? String)!)
+                case RSKeys.Identify.Traits.name: MoEngageSDKAnalytics.sharedInstance.setName((value as? String)!)
+                case RSKeys.Identify.Traits.phone: MoEngageSDKAnalytics.sharedInstance.setMobileNumber(value as! String)
+                case RSKeys.Identify.Traits.firstName: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: "USER_ATTRIBUTE_USER_FIRST_NAME")
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:171]: I can see previously `USER_ATTRIBUTE_USER_FIRST_NAME` was some constant and now it's some string as you've defined it inside `"..."`. Please use the correct key.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 10]
```diff
@@ -137,15 +165,15 @@ extension RSMoEngageDestination {
...
-                case RSKeys.Identify.Traits.address: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.address)
-                case RSKeys.Identify.Traits.age: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.age)
+                case RSKeys.Identify.Traits.email: MoEngageSDKAnalytics.sharedInstance.setEmailID((value as? String)!)
+                case RSKeys.Identify.Traits.name: MoEngageSDKAnalytics.sharedInstance.setName((value as? String)!)
+                case RSKeys.Identify.Traits.phone: MoEngageSDKAnalytics.sharedInstance.setMobileNumber(value as! String)
+                case RSKeys.Identify.Traits.firstName: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: "USER_ATTRIBUTE_USER_FIRST_NAME") 
+                case RSKeys.Identify.Traits.lastName: MoEngageSDKAnalytics.sharedInstance.setLastName((value as? String)!)
+                case RSKeys.Identify.Traits.gender: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: "USER_ATTRIBUTE_USER_GENDER")
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:173]: I can see previously `USER_ATTRIBUTE_USER_GENDER` was some constant and now it's some string as you've defined it inside "...". Please use the correct key.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 11]
```diff
@@ -137,15 +165,15 @@ extension RSMoEngageDestination {
...
+                case RSKeys.Identify.Traits.birthday: identifyDateUserAttribute(value: value, key: "USER_ATTRIBUTE_USER_BDAY")
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:174]: I can see previously `USER_ATTRIBUTE_USER_BDAY` was some constant and now it's some string as you've defined it inside "...". Please use the correct key.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 12]
```diff
@@ -137,15 +165,15 @@ extension RSMoEngageDestination {
...
-                case RSKeys.Identify.Traits.phone: MoEngage.sharedInstance().setUserMobileNo(value)
-                case RSKeys.Identify.Traits.firstName: MoEngage.sharedInstance().setUserAttribute(value, forKey: USER_ATTRIBUTE_USER_FIRST_NAME)
-                case RSKeys.Identify.Traits.lastName: MoEngage.sharedInstance().setUserLastName(value as? String)
-                case RSKeys.Identify.Traits.gender: MoEngage.sharedInstance().setUserAttribute(value, forKey: USER_ATTRIBUTE_USER_GENDER)
-                case RSKeys.Identify.Traits.birthday: identifyDateUserAttribute(value: value, key: USER_ATTRIBUTE_USER_BDAY)
-                case RSKeys.Identify.Traits.address: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.address)
-                case RSKeys.Identify.Traits.age: MoEngage.sharedInstance().setUserAttribute(value, forKey: RSKeys.Identify.Traits.age)
+                case RSKeys.Identify.Traits.email: MoEngageSDKAnalytics.sharedInstance.setEmailID((value as? String)!)
+                case RSKeys.Identify.Traits.name: MoEngageSDKAnalytics.sharedInstance.setName((value as? String)!)
```
Comment (SKannaniOS) [inline Sources/Classes/RSMoEngageDestination.swift:169]: Avoid Force unwrapping the optional value. Use Optional binding.

File: Sources/Classes/RSMoEngageDestination.swift (hunk 9 above)
Comment (SKannaniOS) [inline Sources/Classes/RSMoEngageDestination.swift:171]: Instead of direct string, use enum with string type.

File: Sources/Classes/RSMoEngageDestination.swift (hunk 4 above)
Comment (SKannaniOS) [inline Sources/Classes/RSMoEngageDestination.swift:188]: Use existing method to convert the date to timestamp or create one & use it.

**Review by SKannaniOS — CHANGES_REQUESTED** [review]

File: Sources/Classes/RSMoEngageDestination.swift (hunk 3 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 9 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: Key was not present in earlier project, kindly share if you have any reference for previous key used

File: Sources/Classes/RSMoEngageDestination.swift (hunk 10 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift:185]: I searched for key in previous version. Key was not present in earlier project, kindly share if you have any reference for previous key used

File: Sources/Classes/RSMoEngageDestination.swift (hunk 11 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift:186]: I searched for key in previous version. Key was not present in earlier project, kindly share if you have any reference for previous key used

File: Sources/Classes/RSMoEngageDestination.swift (hunk 12 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 9 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 4 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift [hunk 13]
```diff
@@ -137,15 +167,27 @@ extension RSMoEngageDestination {
...
+                case RSKeys.Identify.Traits.firstName: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: RSKeys.Identify.Traits.firstName)
+                case RSKeys.Identify.Traits.lastName:
+                    if let lastName = value as? String {
+                       MoEngageSDKAnalytics.sharedInstance.setLastName(lastName)
+                    }
+                case RSKeys.Identify.Traits.gender: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: RSKeys.Identify.Traits.gender)
+                case RSKeys.Identify.Traits.birthday: identifyDateUserAttribute(value: value, key: RSKeys.Identify.Traits.birthday)
+                case RSKeys.Identify.Traits.address: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: RSKeys.Identify.Traits.address)
+                case RSKeys.Identify.Traits.age: MoEngageSDKAnalytics.sharedInstance.setUserAttribute(value, withAttributeName: RSKeys.Identify.Traits.age)
                 default: identifyDateUserAttribute(value: value, key: key)
```
Comment (SKannaniOS) [inline Sources/Classes/RSMoEngageDestination.swift:191]: Split into multiple functions as per the value type & apply the optional binding.
e.x: The 'default' case handles the value as date.

**Review by SKannaniOS — CHANGES_REQUESTED** [review]

File: Sources/Classes/RSMoEngageDestination.swift (hunk 11 above)
Comment (1abhishekpandey) [reply Sources/Classes/RSMoEngageDestination.swift]: Please always refer to the destination doc for such references. [Here](https://developers.moengage.com/hc/en-us/articles/4403905883796-Tracking-user-attributes#01GP396BP8DYSAW3MF9W2VMWVB) you go:
```
MoEngageSDKAnalytics.sharedInstance.setDateOfBirth(userBirthdate)//userBirthdate should be a Date instance
```

File: Sources/Classes/RSMoEngageDestination.swift (hunk 10 above)
Comment (1abhishekpandey) [reply Sources/Classes/RSMoEngageDestination.swift]: Please always refer to the destination doc for such references. [Here](https://developers.moengage.com/hc/en-us/articles/4403905883796-Tracking-user-attributes#01GP396BP8DYSAW3MF9W2VMWVB) you go:
```
MoEngageSDKAnalytics.sharedInstance.setGender(.male) //Use UserGender enumerator for this
```
Please set the correct ENUMs (e.g., male/female) in this API.

File: Sources/Classes/RSMoEngageDestination.swift (hunk 4 above)
Comment (1abhishekpandey) [reply Sources/Classes/RSMoEngageDestination.swift]: You need to find the corresponding MoEngage API. As of now, I could see you've used this:
```
MoEngageSDKAnalytics.sharedInstance.setUserAttribute(convertDateToTimestamp(date: convertedDate), withAttributeName: key)
```
This is not appropriate.

Please look at this MoEngage migration doc and you can get more info about the corresponding API: https://developers.moengage.com/hc/en-us/articles/4793185539476-Migration-to-SDK-version-8-2-0#01H8K6YBXX7NBQZR004EAXFR7W.

Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: I have already searched on migration doc before while changing code. No such api found for specific timestamp on migration doc, also existing api removed.

Comment (1abhishekpandey) [reply Sources/Classes/RSMoEngageDestination.swift]: It's here: 
image

File: Sources/Classes/RSMoEngageDestination.swift (hunk 13 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift:191]: Done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 11 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 10 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 9 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift [hunk 14]
```diff
@@ -6,62 +6,83 @@
...
+            moEngageDataCenter = .data_center_02
+        } else if moEngageConfig.region == "US" {
+            moEngageDataCenter = .data_center_01
+        } else if moEngageConfig.region == "IND" {
+            moEngageDataCenter = .data_center_03
+        }
+        else{
+            client?.log(message:"MoEngage SDK initialization terminated due to an invalid region.", logLevel: .warning)
+        }
+
```
[the remark this answers is not in the document]
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 4 above)
Comment (shwetalodhe123) [reply Sources/Classes/RSMoEngageDestination.swift]: done

File: Sources/Classes/RSMoEngageDestination.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply Sources/Classes/RSMoEngageDestination.swift]: You're going to do this change again, as we discussed offline.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 15]
```diff
@@ -6,62 +6,89 @@
...
+        if moEngageConfig.region == "EU" {
+            moEngageDataCenter = .data_center_02
+        } else if moEngageConfig.region == "US" {
+            moEngageDataCenter = .data_center_01
+        } else if moEngageConfig.region == "IND" {
+            moEngageDataCenter = .data_center_03
+        }
+        else{
+            client?.log(message:"MoEngage SDK initialization terminated due to an invalid region.", logLevel: .warning)
+        }
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:39]: As we are terminating, please add ` return` statement after the log statement. Otherwise, it'll continue to initialise the SDK.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 16]
```diff
@@ -6,62 +6,89 @@
...
+       
+        let sdkConfig = MoEngageSDKConfig(appId: moEngageConfig.apiId, dataCenter: moEngageDataCenter!)
+        
+        //enable debugging
+        if client?.configuration?.logLevel != .none {
+            MoEngageCore.MoEngageConsoleLogConfig(isLoggingEnabled: true, loglevel: MoEngageCore.MoEngageLoggerType.debug)
+        }
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:46]: Please follow [this](https://github.com/rudderlabs/rudder-integration-moengage-swift/pull/4/commits/c45dbf7b757f6aa28e3fd95c0184c546855af665) commit to properly implement the MoEngage log level.

As of now, the log level is not even set.

**Review by 1abhishekpandey — APPROVED** [review]

**Review by SKannaniOS — APPROVED** [review]

File: Sources/Classes/RSMoEngageDestination.swift (hunk 16 above)
Comment (1abhishekpandey) [reply Sources/Classes/RSMoEngageDestination.swift]: @shwetalodhe123 I seem to have made a mistake in my commit (which I've referred above). Please follow this updated logic:
```
if client?.configuration?.logLevel != .none {
    sdkConfig.consoleLogConfig = getMoEngageLogLevel(client?.configuration?.logLevel)
}
```

```
func getMoEngageLogLevel(logLevel: RSLogLevel) -> MoEngageCore.MoEngageConsoleLogConfig {
        if (logLevel == RSLogLevel.error) {
            return MoEngageCore.MoEngageConsoleLogConfig(isLoggingEnabled: true, loglevel: MoEngageCore.MoEngageLoggerType.error)
        } else if (logLevel == RSLogLevel.warning) {
            return MoEngageCore.MoEngageConsoleLogConfig(isLoggingEnabled: true, loglevel: MoEngageCore.MoEngageLoggerType.warning)
        } else if (logLevel == RSLogLevel.info) {
            return MoEngageCore.MoEngageConsoleLogConfig(isLoggingEnabled: true, loglevel: MoEngageCore.MoEngageLoggerType.info)
        } else if (logLevel == RSLogLevel.debug) {
            return MoEngageCore.MoEngageConsoleLogConfig(isLoggingEnabled: true, loglevel: MoEngageCore.MoEngageLoggerType.verbose)
        }
    }
```

Please make sure to test this logic.

File: Sources/Classes/RSMoEngageDestination.swift [hunk 17]
```diff
@@ -6,62 +6,96 @@
...
+       
+        let sdkConfig = MoEngageSDKConfig(appId: moEngageConfig.apiId, dataCenter: moEngageDataCenter!)
+        
+        //enable debugging
+        if client?.configuration?.logLevel != .none {
+            sdkConfig.consoleLogConfig = MoEngageCore.MoEngageConsoleLogConfig(isLoggingEnabled: true, loglevel: MoEngageCore.MoEngageLoggerType.debug)
+        }
+        if client?.configuration?.logLevel != .none {
+            sdkConfig.consoleLogConfig = getMoEngageLogLevel(client?.configuration?.logLevel)
+        }
```
Comment (1abhishekpandey) [inline Sources/Classes/RSMoEngageDestination.swift:50]: There are duplicate codes, please remove the code which is not required.

File: Examples/SampleAppSwift/SampleAppSwift/AppDelegate.swift [hunk 18]
```diff
@@ -19,8 +19,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
 
-        let config: RSConfig = RSConfig(writeKey: "27COeQCO3BS2WMw8CJUqYRC5hL7")
-            .dataPlaneURL("https://rudderstacbumvdrexzj.dataplane.rudderstack.com")
+        let config: RSConfig = RSConfig(writeKey: "2k3Dk3s8vs7Y8JPoCfWPQd4Pm54")
+            .dataPlaneURL("https://rudderstacuodq.dataplane.rudderstack.com")
```
Comment (1abhishekpandey) [inline Examples/SampleAppSwift/SampleAppSwift/AppDelegate.swift:23]: Revert this change.

File: Examples/SampleAppObjC/SampleAppObjC/AppDelegate.m [hunk 19]
```diff
@@ -20,9 +21,9 @@ @implementation AppDelegate
 
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
     // Override point for customization after application launch.
-    RSConfig *config = [[RSConfig alloc] initWithWriteKey:@"27COeQCO3BS2WMw8CJUqYRC5hL7"];
-    [config dataPlaneURL:@"https://rudderstacz.dataplane.rudderstack.com"];
-    [config loglevel:RSLogLevelDebug];
+    RSConfig *config = [[RSConfig alloc] initWithWriteKey:@"2k3Dk3s8vs7Y8JPoCfWPQd4Pm54"];
+    [config dataPlaneURL:@"https://rudderstacuodq.dataplane.rudderstack.com"];
```
Comment (1abhishekpandey) [inline Examples/SampleAppObjC/SampleAppObjC/AppDelegate.m:25]: Revert this change.

File: Examples/SampleAppObjC/SampleAppObjC/AppDelegate.m [hunk 20]
```diff
@@ -32,9 +33,23 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     [client addDestination:[[RudderMoEngageDestination alloc] init]];
     [client track:@"Track 1"];
     
-    if (@available(iOS 10.0, *)) {
-        UNUserNotificationCenter.currentNotificationCenter.delegate = self;
-    }
+//    if (@available(iOS 10.0, *)) {
+//        UNUserNotificationCenter.currentNotificationCenter.delegate = self;
+//    }
```
Comment (1abhishekpandey) [inline Examples/SampleAppObjC/SampleAppObjC/AppDelegate.m:38]: If they are not needed please remove them or else un-comment.

File: Examples/SampleAppObjC/SampleAppObjC/AppDelegate.m [hunk 21]
```diff
@@ -43,7 +58,11 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
 
 - (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
     [[RSClient sharedInstance] application:application didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
-
+    NSString *deviceTokenString = [self deviceTokenToString:deviceToken];
+        NSLog(@"Device Token: %@", deviceTokenString);
+    //[[MoEngageSDKMessaging sharedInstance] setPushToken:deviceToken];
```
Comment (1abhishekpandey) [inline Examples/SampleAppObjC/SampleAppObjC/AppDelegate.m:63]: Please remove commented code.

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]
